src/ZODB/tests/racetest.py

```diff
@@ -0,0 +1,515 @@
+        root['obj2'].value += 1
+
+    def assertStateOK(_, root):
+        # both objects must have the same values
+        i1 = root['obj1'].value
+        i2 = root['obj2'].value
+
+        if not (i1 == i2):
+            raise AssertionError("obj1 (%d)  !=  obj2 (%d)" % (i1, i2))
+
+
+class T2ObjectsInc2Phase(ISpec):
+    """T2ObjectsInc2Phase is specification with behaviour where two objects
+    obj1 and obj2 are incremented in lock-step.
+
+    It is used in tests where bugs can be observed on the next transaction
+    after the race.
+
+    invariant:  obj1 - obj2 == phase
+    """
+    def init(_, root):
+        root['obj1'] = MinPO(0)
+        root['obj2'] = MinPO(0)
+        root['phase'] = MinPO(0)
+
+    def next(_, root):
+        phase = root['phase']
+        if phase.value == 0:
+            root['obj1'].value += 1
+        else:
+            root['obj2'].value += 1
+        phase.value += 1
+        phase.value %= 2
+
+    def assertStateOK(_, root):
+        i1 = root['obj1'].value
+        i2 = root['obj2'].value
+        p = root['phase'].value
+
+        if not (i1 - i2 == p):
+            raise AssertionError("obj1 (%d) - obj2(%d) != phase (%d)" %
+                                 (i1, i2, p))
+
+
+class RaceTests(object):
+
+    # verify storage/Connection for race in between load/open and local
+    # invalidations.
+    # https://github.com/zopefoundation/ZEO/issues/166
+    # https://github.com/zopefoundation/ZODB/issues/290
+    def check_race_loadopen_vs_local_invalidate(self):
+        return self._check_race_loadopen_vs_local_invalidate(T2ObjectsInc())
+
+    @with_high_concurrency
+    def _check_race_loadopen_vs_local_invalidate(self, spec):
+        assert isinstance(spec, ISpec)
+        db = DB(self._storage)
+
+        # init initializes the database according to the spec.
+        def init():
+            _state_init(db, spec)
+
+        # verify accesses the database and verifies spec invariant.
+        #
+        # Access to half of the objects is organized to always trigger loading
+        # from zstor. Access to the other half goes through zconn cache and so
+        # verifies whether the cache is not stale.
+        failed = threading.Event()
+        failure = [None]
+
+        def verify():
+            transaction.begin()
+            zconn = db.open()
+            root = zconn.root()
+
+            # reload some objects from zstor, while getting others from
+            # zconn cache
+            _state_invalidate_half1(root)
+
+            try:
+                spec.assertStateOK(root)
+            except AssertionError as e:
+                msg = "verify: %s\n" % e
+                msg += _state_details(root)
+                failure[0] = msg
+                failed.set()
+
+            # we did not changed anything; also fails with commit:
+            transaction.abort()
+            zconn.close()
+
+        # modify changes the database by executing "next" step.
```

Comment:
Similarly to https://github.com/zopefoundation/ZODB/pull/368#discussion_r946364961 the text here, before the genralization, was:

```py
        # modify changes obj1/obj2 by doing `objX.value += 1`.
        #
        # Since both objects start from 0, the invariant that
        # `obj1.value == obj2.value` is always preserved.
        def modify():
           ...
```

So do you think that maybe it might be more uniform with

```py
    # modify changes objects in the database by executing "next" step.
    #
    # Spec invariant should be preserved.
    def modify():
```

?